pytoda/datasets/distributional_dataset.py

```diff
@@ -0,0 +1,112 @@
+from dataclasses import dataclass
+from typing import Optional, Tuple, Union
+
+import torch
+from torch.distributions.distribution import Distribution
+from torch.random import fork_rng
+from torch.utils.data import Dataset
+
+from pytoda.types import Any, Tensor
+
+
+@dataclass
+class StochasticItems:
+    """Sample an item from a distribution on the fly on indexing.
+
+    Args:
+        distribution (Distribution): the distribution to sample from.
+        shape (torch.Size): the desired shape of each item.
+        device (torch.device): device to send the tensor to.
+    """
+
+    distribution: Distribution
+    shape: Union[torch.Size, Tuple[int]]
+    device: torch.device
+
+    def __getitem__(self, index: Any) -> Tensor:
+        """Samples an item.
+
+        Args:
+            index (Any): is ignored.
+
+        Returns:
+            Tensor: sampled from distribution with given shape.
+        """
+        return self.distribution.sample(self.shape).to(self.device)
+
+
+class DistributionalDataset(Dataset):
+    """Generates samples from a specified distribution."""
+
+    def __init__(
+        self,
+        dataset_size: int,
+        item_shape: Tuple[int],
+        distribution_function: Distribution,
```

Comment:
We still use the factory the same way, just not inside the class but before instantiating the class.
@niklexical the docs could still mention the factory, and there could even be a default value using the factory.
`distribution_function: Distribution = DISTRIBUTION_FUNCTION_FACTORY['normal](loc=0.0, scale=1.0)`

This is much more usable because one is not restricted to the distributions we happen to include in the factory.

Reply:
I agree on the usability here! Thus my only point is that the documentation should be improved and I would replace `distribution_function: Distribution,` with `distribution_function: torch.distributions.distribution.Distribution,`  although I admit this is a damn long type name :D

Reply:
ah I see. Makes sense, no reason not to be verbose when it helps!